rust/main/utils/run-locally/src/cosmos/termination_invariants.rs

```diff
@@ -0,0 +1,135 @@
+    .sum::<u32>();
+    if msg_processed_count != messages_expected {
+        log!(
+            "Relayer confirmed {} submitted messages, expected {}",
+            msg_processed_count,
+            messages_expected
+        );
+        return Ok(false);
+    }
+
+    let ending_relayer_balance: f64 =
+        agent_balance_sum(relayer_metrics_port).expect("Failed to get relayer agent balance");
+
+    // Make sure the balance was correctly updated in the metrics.
+    // Ideally, make sure that the difference is >= gas_per_tx * gas_cost, set here:
+    // https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/c2288eb31734ba1f2f997e2c6ecb30176427bc2c/rust/utils/run-locally/src/cosmos/cli.rs#L55
+    // What's stopping this is that the format returned by the `uosmo` balance query is a surprisingly low number (0.000003999999995184)
+    // but then maybe the gas_per_tx is just very low - how can we check that? (maybe by simulating said tx)
+    if starting_relayer_balance <= ending_relayer_balance {
+        log!(
+            "Expected starting relayer balance to be greater than ending relayer balance, but got {} <= {}",
+            starting_relayer_balance,
+            ending_relayer_balance
+        );
+        return Ok(false);
+    }
+
+    let dispatched_messages_scraped = fetch_metric(
+        &scraper_metrics_port.to_string(),
+        "hyperlane_contract_sync_stored_events",
+        &hashmap! {"data_type" => "message_dispatch"},
+    )?
+    .iter()
+    .sum::<u32>();
+    if dispatched_messages_scraped != messages_expected {
+        log!(
+            "Scraper has scraped {} dispatched messages, expected {}",
+            dispatched_messages_scraped,
+            messages_expected
+        );
+        return Ok(false);
+    }
+
+    let gas_payments_scraped = fetch_metric(
+        &scraper_metrics_port.to_string(),
+        "hyperlane_contract_sync_stored_events",
+        &hashmap! {"data_type" => "gas_payment"},
+    )?
+    .iter()
+    .sum::<u32>();
+    if gas_payments_scraped != expected_gas_payments {
+        log!(
+            "Scraper has scraped {} gas payments, expected {}",
+            gas_payments_scraped,
+            expected_gas_payments
+        );
+        return Ok(false);
+    }
+
+    let delivered_messages_scraped = fetch_metric(
+        &scraper_metrics_port.to_string(),
+        "hyperlane_contract_sync_stored_events",
+        &hashmap! {"data_type" => "message_delivery"},
+    )?
+    .iter()
+    .sum::<u32>();
+    if delivered_messages_scraped != messages_expected {
+        log!(
+            "Scraper has scraped {} delivered messages, expected {}",
+            delivered_messages_scraped,
+            messages_expected
+        );
+        return Ok(false);
+    }
+
+    if !request_metric_invariant_met(relayer_metrics_port, messages_expected)? {
+        return Ok(false);
+    }
+
+    log!("Termination invariants have been meet");
+    Ok(true)
+}
+
+pub fn request_metric_invariant_met(
```

Comment:
this logic can be reused in `sealevel/termination_invariants.rs` as well

would also be more descriptive to name it `provider_metrics_invariant_met` imo